lucene/core/src/java/org/apache/lucene/search/QueryVisitor.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.lucene.search;
+
+import java.util.Set;
+import java.util.function.Supplier;
+
+import org.apache.lucene.index.Term;
+import org.apache.lucene.util.automaton.Automaton;
+
+/**
+ * Allows recursion through a query tree
+ *
+ * @see Query#visit(QueryVisitor)
+ */
+public abstract class QueryVisitor {
+
+  /**
+   * Called by leaf queries that match on a specific term
+   *
+   * @param term  the term the query will match on
+   */
+  public void matchesTerm(Term term) { }
+
+  /**
+   * Called by leaf queries that match terms against an Automaton
+   * @param query               the leaf query
+   * @param field               the field to match against
+   * @param isBinary            {@code true} if the Automaton is built to match against byte[] rather than char[]
+   * @param automatonSupplier   a supplier for the built Automaton
+   */
+  public void matchesAutomaton(Query query, String field, boolean isBinary, Supplier<Automaton> automatonSupplier) {}
```

Comment:
The whole Automaton API is still kind of trappy, but I think that's a larger issue.  I think the flag is the least invasive way of doing things for now.

Reply:
I really dont' like that this name implies a question and then it doesn't have a return type. I think we should rather call it `consumeTerm` / `consumeAutomaton` instead. I agree with @dsmiley if we can put this boolean on the Automaton it would be great. There is also the issue of the Supplier which I think is a design smell I am really torn on this method to be honest

Reply:
here is a different suggestion: maybe we can make `FuzzyQuery` subclass `AutomatonQuery` and then simply add a method `visitAutomaton(AutomatonQuery query) { visiteLeaf(query); }` then we can remove this method an users can just pull the automaton, field and if it's binary from `AutomatonQuery`.  This would be a much cleaner API. We don't even have to add an API for this (there is only one users) we can just do an `instanceof` check?

Reply:
++